### PR TITLE
Module 5

### PR DESCRIPTION
**Pull request by MariaMillanGordillo**

### CODE REVIEW

File: src/module_5/module5.ipynb
Comment (pabaldonedo) [inline src/module_5/module5.ipynb:1]: Upload a .md otherwise the diff does not render and I cannot comment

**Review by pabaldonedo — COMMENTED** [review]
Te había dejado un comentario sobre el md pero quedó en borrador hace 5 días! de ahí el retraso de este comentario

File: src/module_5/module5.ipynb
Comment (MariaMillanGordillo) [reply src/module_5/module5.ipynb:1]: Siempre se me olvida, perdona 😅
Ya está subido!

File: src/module_5/module5.md [hunk 1]
```diff
@@ -0,0 +1,1158 @@
...
+```python
+print(f"Given model mean metric: {model_df['model_weighted_return'].mean()}")
+print(f"Baseline model mean metric: {baseline_df['baseline_top5_mktcap'].mean()}")
+```
+
+    Given model mean metric: 0.261664439106911
+    Baseline model mean metric: 0.04567152871752443
+
+
+The baseline model appears to be more stable, however the given model adds more risk, increasing the error.
```
Comment (pabaldonedo) [inline src/module_5/module5.md]: I do not understand what you mean:
 - more stable: by the nature of the problem is expected to make most of the gains in big swings in handful of days, so "stability" is not really a big thing
 - "more risk -> incresing error". We are seeing return no error, here.

Could you explain it further?

File: src/module_5/module5.md [hunk 2]
```diff
@@ -0,0 +1,1158 @@
...
+
+stats = train_results.groupby("n_trees_cat")["l2"].agg(["mean", "std"]).reset_index()
+stats["ymin"] = stats["mean"] - stats["std"]
+stats["ymax"] = stats["mean"] + stats["std"]
+
+plot = (
+    ggplot(stats, aes("n_trees_cat", "mean"))
+    + geom_col(fill="lightblue")
+    + geom_errorbar(aes(ymin="ymin", ymax="ymax"), width=0.2)
+    + labs(x="n_trees", y="binary_logloss (media ± std)", title="binary_logloss por n_trees")
```
Comment (pabaldonedo) [inline src/module_5/module5.md]: mean* +- std

File: src/module_5/module5.md [hunk 3]
```diff
@@ -0,0 +1,1158 @@
...
+    + theme(axis_text_x=element_text(angle=90, vjust=0.5, hjust=1))
+)
+plot
+```
+
+
+
+
+    
+![png](module5_files/module5_41_0.png)
```
Comment (pabaldonedo) [inline src/module_5/module5.md:578]: FYI: std if for 68% of samples assuming Normal distribution, you may want to use 2* for 95% 
https://en.wikipedia.org/wiki/68%E2%80%9395%E2%80%9399.7_rule

File: src/module_5/module5.md [hunk 4]
```diff
@@ -0,0 +1,1158 @@
...
+    + theme(axis_text_x=element_text(angle=90, vjust=0.5, hjust=1))
+)
+plot
+```
+
+
+
+
+    
+![png](module5_files/module5_42_0.png)
```
Comment (pabaldonedo) [inline src/module_5/module5.md:606]: Add in the title if it is train or test for ease of reading

File: src/module_5/module5.md [hunk 5]
```diff
@@ -0,0 +1,1158 @@
...
+    
+
+
+
+This indicates that while the model's binary logloss improves steadily on the training set as the number of trees increases as the curve is decreasing, the logloss on the test set remains almost constant. This pattern suggests possible overfitting: the model fits the training data better with more trees, but does not generalize to the test data.
+
+### Retrain the model
+
+We try with different hyperparameters in order to reduce the overfitting:
+- `learning_rate`: Lowering it makes each boosting step smaller, which helps the model generalize better.
```
Comment (pabaldonedo) [inline src/module_5/module5.md:616]: Why do you say "generalize better"? It makes it more stable and more likely to converge, but also slower.

File: src/module_5/module5.md [hunk 6]
```diff
@@ -0,0 +1,1158 @@
...
+- `n_estimators`: Selecting an optimal number of trees prevents the model from growing too complex, as too many trees can lead to overfitting .
+- `num_leaves`: Limiting the number of leaves per tree restricts the complexity of each individual tree, ensuring the model captures robust patterns rather than noise.
+
+
+```python
+params = {
+    "random_state": 1,
```
Comment (pabaldonedo) [inline src/module_5/module5.md]: Where do these numbers come from? Are you just checking one set of params?

File: src/module_5/module5.md [hunk 7]
```diff
@@ -0,0 +1,1158 @@
...
+
+
+```python
+print(f"Hyperparametrized model mean metric: {test_results_final_tree['weighted-return'].mean()}")
+```
+
+    Hyperparametrized model mean metric: 0.10401870245714662
+
+
+It improves our given model. However, it is not close to the baseline one.
```
Comment (pabaldonedo) [inline src/module_5/module5.md]: If we are seeing weighted-return, we want the return to be maximum, why do you say this is better?

File: src/module_5/module5.md [hunk 8]
```diff
@@ -0,0 +1,1158 @@
...
+
+
+
+    
+![png](module5_files/module5_51_0.png)
+    
+
+
+
+Now we can appreciate a little improvement, since our data in the plot is decreasing.
```
Comment (pabaldonedo) [inline src/module_5/module5.md]: in the test it is not the case, since all the errors are within std

File: src/module_5/module5.md [hunk 9]
```diff
@@ -0,0 +1,1158 @@
...
+
+
+    
+![png](module5_files/module5_57_0.png)
+    
+
+
+
+
+```python
```
Comment (pabaldonedo) [inline src/module_5/module5.md]: Is not this cell same as previous one but with error bars?

File: src/module_5/module5.md [hunk 10]
```diff
@@ -0,0 +1,1158 @@
...
+import pandas as pd
+from plotnine import ggplot, aes, geom_col, coord_flip, scale_x_discrete, geom_errorbar
```
Comment (pabaldonedo) [inline src/module_5/module5.md]: Imports on top of notebook

File: src/module_5/module5.md [hunk 11]
```diff
@@ -0,0 +1,1158 @@
...
+stats_top = stats.nlargest(20, 'mean').sort_values('mean')
+
+plot = (
+    ggplot(stats_top, aes(x='feature', y='mean'))
+    + geom_col(fill='lightblue')
+    + geom_errorbar(aes(ymin=stats_top['mean']-stats_top['var']**0.5, ymax=stats_top['mean']+stats_top['var']**0.5), width=0.2)
+    + coord_flip()
+    + scale_x_discrete(limits=stats_top['feature'])
+)
+plot
```
Comment (pabaldonedo) [inline src/module_5/module5.md:876]: Why do you want to merge all models into one? I think that can confuse more than help

File: src/module_5/module5.md [hunk 12]
```diff
@@ -0,0 +1,1158 @@
...
+
+    # summary plot (importancia global para este test set)
+    plt.figure(figsize=(16, 10))
+    shap.summary_plot(sv, X, show=False)
+    plt.tight_layout()
+    display(plt.gcf())
+    plt.clf()
+
+if len(all_models) > 0:
+    fecha_ejemplo = sorted(all_models.keys())[-1]
```
Comment (pabaldonedo) [inline src/module_5/module5.md]: sample_date

**Review by pabaldonedo — COMMENTED** [review]
In general good work.

I have left some comments and questions. The most shocking for me is why did not you try more hyperparameterizations?

Also, I am not commenting things that Guillermo covered in his class (like data leakage in the dataset because of splits in shares, and so on)

File: src/module_5/module5.md [hunk 13]
```diff
@@ -617,52 +619,88 @@ We try with different hyperparameters in order to reduce the overfitting:
...
+        vals = valid[metric]
+        try:
+            last = vals[-1]
+        except Exception:
+            last = list(vals)[-1]
+        losses.append(last)
+    return float(np.mean(losses)) if len(losses) > 0 else float("inf")
+
+for _ in range(5):
+    p = sample_params()
```
Comment (pabaldonedo) [inline src/module_5/module5.md:664]: In case you are no aware of you may be interested in this https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.RandomizedSearchCV.html
